iroh-net/src/magicsock/derp_actor.rs

```diff
@@ -49,33 +48,108 @@ pub(super) enum DerpActorMessage {
 #[derive(Debug)]
 struct ActiveDerp {
     c: derp::http::Client,
-    cancel: CancellationToken,
     /// The time of the last request for its write
     /// channel (currently even if there was no write).
     last_write: Instant,
-    create_time: Instant,
     reader: ReaderState,
+    inbox: mpsc::Receiver<ActiveDerpMessage>,
+    msg_sender: mpsc::Sender<ActorMessage>,
+    /// Contains optional alternate routes to use as an optimization instead of
+    /// contacting a peer via their home DERP connection. If they sent us a message
+    /// on this DERP connection (which should really only be on our DERP
+    /// home connection, or what was once our home), then we remember that route here to optimistically
+    /// use instead of creating a new DERP connection back to their home.
+    derp_routes: Vec<PublicKey>,
 }
 
-/// A route entry for a public key, saying that a certain peer should be available at DERP
-/// node derpID, as long as the current connection for that derpID is dc. (but dc should not be
-/// used to write directly; it's owned by the read/write loops)
 #[derive(Debug)]
-struct DerpRoute {
-    derp_id: u16,
-    dc: derp::http::Client, // don't use directly; see comment above
+#[allow(clippy::large_enum_variant)]
+enum ActiveDerpMessage {
+    GetLastWrite(oneshot::Sender<Instant>),
+    Ping(oneshot::Sender<Result<Duration, ClientError>>),
+    GetLocalAddr(oneshot::Sender<Option<SocketAddr>>),
+    GetPeerRoute(PublicKey, oneshot::Sender<Option<derp::http::Client>>),
+    GetClient(oneshot::Sender<derp::http::Client>),
+    NotePreferred(bool),
+    Shutdown,
+}
+
+impl ActiveDerp {
+    async fn run(mut self) {
+        if let Err(err) = self.c.connect().await {
+            // TODO: what to do?
```

Comment:
if the error if `ClientError::Timeout`, probably just do a backoff and try to dial again later?

The other option is to just shut down the active derp and remove it from the `DerpActor` & then wait until we need to dial that derper again

Reply:
fixed